### PR TITLE
Docker add build matrix

### PR DESCRIPTION
**Pull request by maryapp**
Here is the updates to merge the main PR after this one.
Re: @MMathisLab

Main minor changes:

- delete init of args
- version of cuda and dlc

Last - new images on dockerhub are functional and built from this version of code!

PS sorry for "shift reds" in build.sh, probably we are using different style-formatter + I added some minor docs-lines

### CODE REVIEW

File: docker/Dockerfile.base [hunk 1]
```diff
@@ -1,25 +1,29 @@
...
+ARG CUDA_VERSION
 FROM nvidia/cuda:${CUDA_VERSION}
 
-ARG DEEPLABCUT_VERSION=2.2.1.1
+ARG DEEPLABCUT_VERSION
 ENV DEBIAN_FRONTEND=noninteractive
 
 RUN apt-get update -yy \ 
-    && apt-get install -yy --no-install-recommends python3 python3-pip python3-dev ffmpeg libsm6 libxext6 \
+    && apt-get install -yy --no-install-recommends python3.8 python3-pip python3-dev ffmpeg libsm6 libxext6 \
```
Comment (jeylau) [inline docker/Dockerfile.base]: Why not a more recent python version?

Comment (maryapp) [reply docker/Dockerfile.base]: from docs, with numpy 1.21, the python 3.9 is our max, python 3.8 or 3.9 not really a big difference. However, if you have another working combination for all dependencies, you are welcome to propose it in the last dlc-docker dedicated post!

**Comment by MMathisLab** [comment]
3.9 is better as there are something, like cebra, that don't work in 3.8 and we want to try to keep interoperability :D

**Comment by maryapp** [comment]
> 3.9 is better as there are something, like cebra, that don't work in 3.8 and we want to try to keep interoperability :D

ok, no pb, I can rebuild for dockerhub, I guess we were discussing different dependencies preferences, but not the python one haha

**Comment by MMathisLab** [comment]
@stes LGTY?

**Comment by maryapp** [comment]
@MMathisLab we discussed with @jeylau to delete python3-dev and protobuff from Dockerfile: I need to rebuild and test again, but there is this "mirror bug" still, so need to wait (Monday?)

**Comment by jeylau** [comment]
@maryapp to "try" removing python3-dev to see if that explains python3.8 and python3.9 living together in the container 😅

**Comment by maryapp** [comment]
updates about python3.9 situation: it was python3-pip that was polluting the python paths, and running dlc still under 3.8. I addressed paths, but under python3.9, numpy wakes up with bugs and incompatibilities. (TypeError: 'numpy._DTypeMeta' object is not subscriptable)
@jeylau you could share with me your env for sure, but it's not unfer TF 2.5, so I don't think it will help. 
Do we really want python3.9 now? Feels like not smooth and quick fix...

File: docker/Dockerfile.base (hunk 1 above)
Comment (maryapp) [reply docker/Dockerfile.base]: PS TF 2.5 asks for numpy ~1.19, under python3.8 the numpy is 1.21.6 works: any upgrade creates errors during the dlc imput ("AttributeError: module 'numpy' has no attribute 'object'" or another0" module 'numpy' has no attribute 'typeDict'). So it doesn't seem we are able to find a better working configuration.

Comment (MMathisLab) [reply docker/Dockerfile.base]: Our deployed conda env is 3.9; there should be no need to pin numpy in the Docker, it's set in dlc setup.py ?

Comment (maryapp) [reply docker/Dockerfile.base]: I was following the ensemble of dependencies that have already existed  in the initial Dockerfile, and eventually I would like to get rid off of any duplicates (from setup etc), so once we have a one stable version, I was testing, what could we throw away. Concerning numpy, I think that the version of numpy is precised explicitly, as in the dlc setup.py it's >=1.18.5, however how we can notice, it's working only with 1.21 without any dramatic conflicts from tensorflow for example. @jeylau what do you think

Comment (maryapp) [reply docker/Dockerfile.base]: Yes, it's the case: once I don't precise the version it installs 1.24.4 and during the dlc import the error "AttributeError: module 'numpy' has no attribute 'object'." appears. (it's not dlc itself, but TensorFlow 2.5 side), but as was discussed earlier we can't upgrade the tf version, as cuda (11.4)

**Comment by jeylau** [comment]
Why can't we update TF? Looking at https://www.tensorflow.org/install/source#gpu, it looks like we could install TF up to 2.11 w/ CUDA 11.2 and python >3.7,<3.10.

**Comment by MMathisLab** [comment]
yes, we should only be using **TF 2.10** -- that is the developer agreed final version of TF we will support.

**Comment by maryapp** [comment]
from what I remember there was an incompatibility between nvidia container (11.4) version and tensorflow (>2.5). But I will try, again! probably it was a different bug so far.

**Comment by maryapp** [comment]
@jeylau can you please give you feedback about the[ old file from main](https://github.com/maryapp/DeepLabCut_matrix_patch/blob/docker-add-build-matrix/testscript_cli.py) (~2yo) that have been used for docker test and still in use (for functional side of dlc in docker container), does it pass the test in your current working env? because with TF 2.10 and recent numpy (auto version 1.24 from config (aka latest)), it fails, even if just the imports of dlc/tf don't show any problems (except warnings I was referencing you in pm))

but with TF 2.10 and numpy 1.21.6 test are ok! So what's the mystery about numpy or test file

**Comment by maryapp** [comment]
@jeylau check it please, now it should be fine!

**Comment by jeylau** [comment]
`2.3.5-base-cuda11.4.3-cudnn8-runtime-ubuntu20.04-latest` looks all good @maryapp 😊

**Comment by maryapp** [comment]
@stes let's merge, please!

File: docker/Dockerfile.base [hunk 2]
```diff
@@ -1,28 +1,34 @@
...
+		torch==1.12 \
+	&& pip3 list
 
 # The installed tensorflow version will not work with the latest protocol buffer version,
 # hence we are fixing the version to 3.20.
-#
 # See https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates
-# for details on why this is needed.
-RUN pip3 install --no-cache-dir protobuf==3.20.1
+# for details on why this is needed. (re: Aug 21: retested, still required)
```
Comment (stes) [inline docker/Dockerfile.base]:
```suggestion
# for details on why this is needed. (re: Aug 21, 2023: retested, still required)
```

File: docker/build.sh
Comment (stes) [inline docker/build.sh:1]: I think this is just whitespace changes and no actual updates, right? can we revert/minimize diff?

**Review by stes — CHANGES_REQUESTED** [review]
minor comments.

I assume this is tested for GUI and console?

File: docker/build.sh
Comment (maryapp) [reply docker/build.sh:1]: with indent 4, it seems to be better, or we need to retrieve you bash formatter somewhere

File: docker/Dockerfile.gui [hunk 3]
```diff
@@ -1,21 +1,17 @@
-ARG CUDA_VERSION=11.4.0-cudnn8-runtime-ubuntu20.04
-ARG DEEPLABCUT_VERSION=2.2.1.1
-FROM deeplabcut/deeplabcut:${DEEPLABCUT_VERSION}-base-cuda${CUDA_VERSION}
+ARG CUDA_VERSION
+ARG DEEPLABCUT_VERSION
+
+FROM deeplabcut/deeplabcut:${DEEPLABCUT_VERSION}-base-cuda${CUDA_VERSION}-latest
```
Comment (stes) [inline docker/Dockerfile.gui:6]: Let's a deprecation notice to this dockerfile given it is currently not pushed (as a comment)

```suggestion
# NOTE: This dockerfile is currently not included in the build process
# It is still left for reference, but currently untested.
ARG CUDA_VERSION
ARG DEEPLABCUT_VERSION

FROM deeplabcut/deeplabcut:${DEEPLABCUT_VERSION}-base-cuda${CUDA_VERSION}-latest
```

**Review by stes — APPROVED** [review]
Assuming that all tests pass, this looks good to me!

**Pull request merged**